pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,363 @@
+	if sess == nil {
+		return nil
+	}
+	// pricing API doesn't have an endpoint in all regions
+	pricingAPIRegion := "us-east-1"
+	if strings.HasPrefix(region, "cn-") {
+		pricingAPIRegion = "cn-north-1"
+	} else if strings.HasPrefix(region, "ap-") {
+		pricingAPIRegion = "ap-south-1"
+	}
+	return pricing.New(sess, &aws.Config{Region: aws.String(pricingAPIRegion)})
+}
+
+func NewPricingProvider(ctx context.Context, pricing pricingiface.PricingAPI, ec2Api ec2iface.EC2API, region string) *PricingProvider {
+	p := &PricingProvider{
+		region:             region,
+		onDemandUpdateTime: initialPriceUpdate,
+		onDemandPrices:     initialOnDemandPrices,
+		spotUpdateTime:     initialPriceUpdate,
+		// default our spot pricing to the same as the on-demand pricing until a price update
+		spotPrices: initialOnDemandPrices,
+		ec2:        ec2Api,
+		pricing:    pricing,
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("pricing"))
+
+	// perform an initial price update at startup
+	p.updatePricing(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				logging.FromContext(ctx).Infof("shutting down")
+				return
+			case <-time.After(pricingUpdatePeriod):
+				p.updatePricing(ctx)
+			}
+		}
+	}()
+	return p
+}
+
+// InstanceTypes returns the list of all instance types for which either a spot or on-demand price is known.
+func (p *PricingProvider) InstanceTypes() []string {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return lo.Union(lo.Keys(p.onDemandPrices), lo.Keys(p.spotPrices))
+}
+
+// OnDemandPrice returns the last known on-demand price for a given instance type, returning an error if there is no
+// known on-demand pricing for the instance type.
+func (p *PricingProvider) OnDemandPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	price, ok := p.onDemandPrices[instanceType]
+	if !ok {
+		return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+	}
+	return price, nil
+}
+
+// SpotPrice returns the last known spot price for a given instance type, returning an error if there is no
+// known spot pricing for the instance type.
+func (p *PricingProvider) SpotPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if price, ok := p.spotPrices[instanceType]; ok {
+		return price, nil
+	}
+	// if there is no spot price available, fall back to the on-demand price
+	if price, ok := p.onDemandPrices[instanceType]; ok {
+		return price, nil
+	}
+	return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+}
+
+func (p *PricingProvider) updatePricing(ctx context.Context) {
+	logging.FromContext(ctx).Infof("Updating EC2 pricing information")
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := p.updateOnDemandPricing(ctx); err != nil {
+			logging.FromContext(ctx).Errorf("updating on-demand pricing, %s, using existing pricing data from %s", err, p.onDemandUpdateTime.Format(time.RFC822Z))
```

Comment:
We don't expect this to work in fully private subnets right? (no IGW or NATGW). I wonder if we'll end up spamming logs in those controllers. 

I like that we're failing open, but should we have a way to turn this off?

Reply:
It's at startup and once every 12 hours so the logs shouldn't be excessive.  If there's some other feature that doesn't work in non-connected Karpenter, maybe add an environment variable then?